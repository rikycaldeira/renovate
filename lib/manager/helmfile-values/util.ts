import { hasKey } from '../../util/object';
import { regEx } from '../../util/regex';
import { HelmfileDockerImageDependency } from './types';

const parentKeyRe = regEx(/image$/i);
const parentKeyReMulti = regEx(/images$/i);

/**
 * Type guard to determine whether a given Helmfile values section potentially
 * defines a Helm Docker dependency in its in-line values
 *
 * There is no exact standard of how Docker dependencies are defined in Helm
 * values, this function defines a
 * heuristic based on the most commonly used format in the Helm charts:
 *
 * image:
 *   repository: 'something'
 *   tag: v1.0.0
 */
export function matchesHelmfileValuesDockerHeuristic(
  parentKey: string,
  data: unknown
): data is HelmfileDockerImageDependency | string {
  return (
    parentKeyRe.test(parentKey) &&
    data &&
    ((typeof data === 'object' &&
      hasKey('repository', data) &&
      hasKey('tag', data)) ||
      typeof data === 'string')
  );
}

/**
 * Type guard to determine whether a given Helmfile values section potentially
 * defines multiple Helm Docker dependencies in its in-line values
 *
 * There is no exact standard of how Docker dependencies are defined in Helm
 * values, this function defines a
 * heuristic to catch the following multi-image format:
 *
 * images:
 *   <image-1>:
 *     repository: 'foo'
 *     tag: v1.0.0
 *   <image-2>:
 *     repository: 'bar'
 *     tag: v2.0.0
 *   <image-3>: 'registry.com/foo:v1.0.0'
 */
export function matchesHelmfileValuesDockerHeuristicMulti(
  parentKey: string,
  data: unknown
): data is HelmfileDockerImageDependency | string {
  return (
    parentKeyReMulti.test(parentKey) &&
    data &&
    typeof data === 'object' &&
    Object.keys(data).every((imageName) => {
      const image = data[imageName];
      return (
        image &&
        ((typeof image === 'object' &&
          hasKey('repository', image) &&
          hasKey('tag', image)) ||
          typeof image === 'string')
      );
    })
  );
}
