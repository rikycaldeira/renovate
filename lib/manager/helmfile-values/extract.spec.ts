import { loadFixture } from '../../../test/util';
import { extractPackageFile } from '.';

const multidocYaml = loadFixture('multidoc.yaml');
const fileName = 'helmfile.yaml';

describe('manager/helmfile-values/extract', () => {
  describe('extractPackageFile()', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    it('returns null if no sections with values', () => {
      const content = `
      releases:
        - name: testRelease
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toBeNull();
    });

    it('do not crash on invalid helmfile.yaml', () => {
      const content = `
      releases: [
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toBeNull();
    });

    it('parse release image', () => {
      const content = `
      releases:
        - name: testRelease
          values:
            - image:
                registry: container.registry.com
                repository: some/repository
                tag: 0.0.1
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/some/repository',
            currentValue: '0.0.1',
          },
        ],
      });
    });

    it('parse release multi images', () => {
      const content = `
      releases:
        - name: testRelease
          values:
            - images:
                foo:
                  registry: container.registry.com
                  repository: some/repository
                  tag: 0.0.1
                bar:
                  repository: another/repository
                  tag: 0.1.0
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/some/repository',
            currentValue: '0.0.1',
          },
          { depName: 'another/repository', currentValue: '0.1.0' },
        ],
      });
    });

    it('parse inline image', () => {
      const content = `
      releases:
        - name: testRelease
          values:
            - image: container.registry.com/some/repository:0.0.1
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/some/repository',
            currentValue: '0.0.1',
          },
        ],
      });
    });

    it('parse inline multi images', () => {
      const content = `
      releases:
        - name: testRelease
          values:
            - images:
                foo: container.registry.com/some/repository:0.0.1
                bar: another/repository:0.1.0
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/some/repository',
            currentValue: '0.0.1',
          },
          { depName: 'another/repository', currentValue: '0.1.0' },
        ],
      });
    });

    it('parse nested state multi images', () => {
      const content = `
      helmfiles:
        - path: another/helmfile.yaml
          values:
            - images:
                foo:
                  registry: container.registry.com
                  repository: some/repository
                  tag: 0.0.1
                bar:
                  repository: another/repository
                  tag: 0.1.0
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/some/repository',
            currentValue: '0.0.1',
          },
          { depName: 'another/repository', currentValue: '0.1.0' },
        ],
      });
    });

    it('parse helmfile with all sections', () => {
      const content = `
      environments:
        default:
          values:
            foo:
              registry: container.registry.com
              repository: some/repository
              tag: 0.0.1

      helmfiles:
        - path: another/helmfile.yaml
          values:
            - images:
                bar:
                  repository: another/repository
                  tag: 0.1.0
      releases:
        - name: testRelease
          values:
            - images:
                bar:
                  registry: container.registry.com
                  repository: yetanother/repository
                  tag: 1.0.0
      `;
      const result = extractPackageFile(content, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          {
            depName: 'container.registry.com/yetanother/repository',
            currentValue: '1.0.0',
          },
          { depName: 'another/repository', currentValue: '0.1.0' },
          // { depName: 'container.registry.com/some/repository', currentValue: '0.0.1' },
        ],
      });
    });

    it('parses multidoc yaml', () => {
      const result = extractPackageFile(multidocYaml, fileName);
      expect(result).toMatchSnapshot({
        datasource: 'docker',
        deps: [
          // { depName: 'container.registry.com/some/repository', currentValue: '0.0.1' },
          { depName: 'another/repository', currentValue: '0.1.0' },
          {
            depName: 'container.registry.com/yetanother/repository',
            currentValue: '1.0.0',
          },
        ],
      });
    });
  });
});
