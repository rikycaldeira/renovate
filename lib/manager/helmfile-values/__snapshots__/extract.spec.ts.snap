// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/helmfile-values/extract extractPackageFile() parse helmfile with all sections 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "1.0.0",
      "datasource": "docker",
      "depName": "container.registry.com/yetanother/repository",
      "replaceString": "1.0.0",
      "versioning": "docker",
    },
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.1.0",
      "datasource": "docker",
      "depName": "another/repository",
      "replaceString": "0.1.0",
      "versioning": "docker",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parse inline image 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.0.1",
      "datasource": "docker",
      "depName": "container.registry.com/some/repository",
      "replaceString": "container.registry.com/some/repository:0.0.1",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parse inline multi images 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.0.1",
      "datasource": "docker",
      "depName": "container.registry.com/some/repository",
      "replaceString": "container.registry.com/some/repository:0.0.1",
    },
    Object {
      "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.1.0",
      "datasource": "docker",
      "depName": "another/repository",
      "replaceString": "another/repository:0.1.0",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parse nested state multi images 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.0.1",
      "datasource": "docker",
      "depName": "container.registry.com/some/repository",
      "replaceString": "0.0.1",
      "versioning": "docker",
    },
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.1.0",
      "datasource": "docker",
      "depName": "another/repository",
      "replaceString": "0.1.0",
      "versioning": "docker",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parse release image 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.0.1",
      "datasource": "docker",
      "depName": "container.registry.com/some/repository",
      "replaceString": "0.0.1",
      "versioning": "docker",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parse release multi images 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.0.1",
      "datasource": "docker",
      "depName": "container.registry.com/some/repository",
      "replaceString": "0.0.1",
      "versioning": "docker",
    },
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.1.0",
      "datasource": "docker",
      "depName": "another/repository",
      "replaceString": "0.1.0",
      "versioning": "docker",
    },
  ],
}
`;

exports[`manager/helmfile-values/extract extractPackageFile() parses multidoc yaml 1`] = `
Object {
  "datasource": "docker",
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "0.1.0",
      "datasource": "docker",
      "depName": "another/repository",
      "replaceString": "0.1.0",
      "versioning": "docker",
    },
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "1.0.0",
      "datasource": "docker",
      "depName": "container.registry.com/yetanother/repository",
      "replaceString": "1.0.0",
      "versioning": "docker",
    },
  ],
}
`;
